packages/lesswrong/server/repos/UltraFeedEventsRepo.ts

```diff
@@ -125,6 +167,251 @@ class UltraFeedEventsRepo extends AbstractRepo<'UltraFeedEvents'> {
 
     return new Set(servedComments.map((row) => row.documentId));
   }
+
+  /**
+   * Fetch a chronological feed of items that were served to the user, grouped as:
+   * - Posts: grouped by post ID (latest serve time)
+   * - Spotlights: grouped by spotlight ID (latest serve time)
+   * - Comment Threads: grouped by (sessionId, itemIndex) to reconstruct a single feed card
+   *
+   * Results are ordered by last served time descending and paginated by cutoff/offset/limit.
+   */
+  async getServedHistoryItems({
+    userId,
+    cutoff,
+    offset = 0,
+    limit = 20,
+  }: {
+    userId: string,
+    cutoff?: Date | null,
+    offset?: number,
+    limit?: number,
+  }): Promise<ServedHistoryItem[]> {
+    const rows = await this.getRawDb().manyOrNone<ServedHistoryItemRaw>(`
+      -- UltraFeedEventsRepo.getServedHistoryItems
+      WITH base_served AS (
+        SELECT
+          ue._id,
+          ue."documentId",
+          ue."collectionName",
+          ue."createdAt",
+          ue.event,
+          ue."userId"
+        FROM "UltraFeedEvents" ue
+        WHERE ue."userId" = $(userId)
+          AND ue."eventType" = 'served'
+          AND ($(cutoff) IS NULL OR ue."createdAt" < $(cutoff))
```

Comment:
Is this fetching every single item served to a given user?

Reply:
They are served events, so they're limited to the last 48 hours, so can't be that many. I put in a LIMIT 5000 though

Reply:
oh right.  hm...